packages/@aws-cdk/aws-iam/lib/grant.ts

```diff
@@ -0,0 +1,257 @@
+   *
+   * (Depending on the resource type, this needs to be '*' in a resource policy).
+   *
+   * @default Same as regular resource ARNs
+   */
+  resourceSelfArns?: string[];
+}
+
+export class Permissions {
+  /**
+   * Grant the given permissions to the principal
+   *
+   * The permissions will be added to the principal policy primarily, falling
+   * back to the resource policy if necessary. The permissions must be granted
+   * somewhere.
+   *
+   * - Trying to grant permissions to a principal that does not admit adding to
+   *   the principal policy while not providing a resource with a resource policy
+   *   is an error.
+   * - Trying to grant permissions to an absent principal (possible in the
+   *   case of imported resources) leads to a warning being added to the
+   *   resource construct.
+   */
+  public static grantWithResource(options: GrantWithResourceOptions): Grant {
+    const result = Permissions.grantOnPrincipal({
+      ...options,
+      scope: options.resource
+    });
+
+    if (result.success) { return result; }
+
+    const statement = new PolicyStatement()
+      .addActions(...options.actions)
+      .addResources(...(options.resourceSelfArns || options.resourceArns))
+      .addPrincipal(options.principal!);
+
+    options.resource.addToResourcePolicy(statement);
+
+    return grantResult({ addedToResource: true, statement, options });
+  }
+
+  /**
+   * Try to grant the given permissions to the given principal
+   *
+   * Absence of a principal leads to a warning, but failing to add
+   * the permissions to a present principal is not an error.
+   */
+  public static grantOnPrincipal(options: GrantOnPrincipalOptions): Grant {
+    if (!options.principal) {
+      if (options.scope) {
+        // tslint:disable-next-line:max-line-length
+        options.scope.node.addWarning(`Could not add grant for '${options.actions}' on '${options.resourceArns}' because the principal was not available. Add the permissions by hand.`);
+      }
+      return grantResult({ principalMissing: true, options });
+    }
+
+    const statement = new PolicyStatement()
+      .addActions(...options.actions)
+      .addResources(...options.resourceArns);
+
+    const addedToPrincipal = options.principal.addToPolicy(statement);
+
+    return grantResult({ addedToPrincipal, statement: addedToPrincipal ? statement : undefined, options });
+  }
+
+  /**
+   * Add a grant both on the principal and on the resource
+   *
+   * As long as any principal is given, granting on the pricipal may fail (in
+   * case of a non-identity principal), but granting on the resource will
+   * always be done.
+   *
+   * Statement will be the resource statement.
+   */
+  public static grantOnPrincipalAndResource(options: GrantOnPrincipalAndResourceOptions): Grant {
+    const result = Permissions.grantOnPrincipal({
+      ...options,
+      scope: options.resource,
+    });
+
+    if (result.principalMissing) { return result; }
```

Comment:
Doesn't that contradict the documentation. I was under the impression that we always update the resource policy in this case

Reply:
Well, if there was something to do. Otherwise, no principal, so what would we add a grant to?

I'm actually a little mistified by why we had `principal?` allowed to be absent again, and what the semantics of that were.